### PR TITLE
BIP 158: Update filter construction

### PR DESCRIPTION
**Pull request by jimpo**
Various things to update implementation to match latest revision of [BIP 158](https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki).

### CODE REVIEW

File: gcs/builder/builder.go [hunk 1]
```diff
@@ -217,7 +221,16 @@ func (b *GCSBuilder) Build() (*gcs.Filter, error) {
 		return nil, b.err
 	}
 
-	return gcs.BuildGCSFilter(b.p, b.key, b.data)
+	dataSlice := make([][]byte, 0, len(b.data))
+	for encoded := range b.data {
+		item, err := base64.StdEncoding.DecodeString(encoded)
```
Comment (Roasbeef) [inline gcs/builder/builder.go]: I think we can actually just do `string(encoded)` here. We don't need the encoding outside this function at all (doesn't need to be human readable), and IIRC there're fast paths in the compiler for things like `map[string(kek)] = 9`.

File: gcs/gcs.go [hunk 2]
```diff
@@ -98,7 +102,7 @@ func BuildGCSFilter(P uint8, key [KeySize]byte, data [][]byte) (*Filter, error)
 	if len(data) == 0 {
 		return nil, ErrNoData
 	}
-	if len(data) > math.MaxInt32 {
+	if len(data) > math.MaxUint32 {
```
Comment (Roasbeef) [inline gcs/gcs.go]: IIRC we changed this as we had issues on 32-bit arm. I'll check to see if this compiles properly on my pi.

File: gcs/gcs.go [hunk 3]
```diff
@@ -372,15 +429,15 @@ func (f *Filter) MatchAny(key [KeySize]byte, data [][]byte) (bool, error) {
...
+	var quotient uint64
 
 	// Count the 1s until we reach a 0.
 	c, err := b.ReadBit()
 	if err != nil {
 		return 0, err
 	}
 	for c {
-		v++
+		quotient++
```
Comment (Roasbeef) [inline gcs/gcs.go:438]: 👍

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: gcs/gcs.go (hunk 2 above)
Comment (jimpo) [reply gcs/gcs.go]: You might be right -- I'll rewrite this line to be safe.

File: gcs/builder/builder.go (hunk 1 above)
Comment (jimpo) [reply gcs/builder/builder.go]: Done.

**Review by Roasbeef — APPROVED** [review]
LGTM 🎉

Won't merge this quite yet until I update neutrino to support the new p2p modifications. As otherwise, people will fetch my fork and find that it doesn't actually work with their node...

**Pull request merged**